storage/points_writer.go

```diff
@@ -8,13 +8,42 @@ import (
 	"github.com/influxdata/influxdb/v2"
 	"github.com/influxdata/influxdb/v2/kit/platform"
 	"github.com/influxdata/influxdb/v2/models"
+	"go.uber.org/zap"
 )
 
 // PointsWriter describes the ability to write points into a storage engine.
 type PointsWriter interface {
 	WritePoints(ctx context.Context, orgID platform.ID, bucketID platform.ID, points []models.Point) error
 }
 
+// MirroringPointsWriter sends all incoming points to both a Primary and Secondary writer.
+// Failures to write points to the Primary writer are bubbled up as errors. Failures to write
+// to the Secondary writer are logged, but not returned.
+type MirroringPointsWriter struct {
+	Primary   PointsWriter
+	Secondary PointsWriter
+	Log       *zap.Logger
+}
+
+func (w *MirroringPointsWriter) WritePoints(ctx context.Context, orgID platform.ID, bucketID platform.ID, p []models.Point) error {
+	if len(p) == 0 {
+		return nil
+	}
+
+	// Try writing to the primary first. If it fails, don't attempt to mirror to the secondary.
+	if err := w.Primary.WritePoints(ctx, orgID, bucketID, p); err != nil {
+		return err
+	}
+
+	// Mirror points to the secondary writer, logging on failure.
+	if err := w.Secondary.WritePoints(ctx, orgID, bucketID, p); err != nil {
```

Comment:
I wonder if it'd be worth doing this async in a goroutine...

Reply:
If you do, you should move it before the `Primary.WritePoints()`, so it doesn't wait for that to complete.

Reply:
That would potentially allow the secondary write to succeed when the primary fails. For the intended use-case of primary=local-TSM and secondary=replication-system, that would be confusing behavior IMO

Reply:
Decision: run this async in a goroutine, after confirming the write to local TSM succeeded.